src/hotspot/share/oops/compressedKlass.cpp

```diff
@@ -90,19 +90,19 @@ void CompressedKlassPointers::initialize(address addr, size_t len) {
   set_shift(shift);
   set_range(range);
 
-  assert(is_valid_base(_base), "Address must be a valid encoding base");
+  assert(is_valid_base(_base), "Address must be a valid encoding base (" PTR_FORMAT ")", p2i(_base));
 }
 
 // Given an address p, return true if p can be used as an encoding base.
 //  (Some platforms have restrictions of what constitutes a valid base address).
 bool CompressedKlassPointers::is_valid_base(address p) {
 #ifdef AARCH64
-  // Below 32G, base must be aligned to 4G.
-  // Above that point, base must be aligned to 32G
-  if (p < (address)(32 * G)) {
-    return is_aligned(p, 4 * G);
-  }
-  return is_aligned(p, (4 << LogKlassAlignmentInBytes) * G);
+  // Any base that is 4G aligned (including null) can probably be made to work.
+  // Without knowing the location and size of the range-to-be-encoded in relation
+  //  to the base it is not possible to say more. If the base is not 0, we assume
+  //  the range starts at the specified encoding base and its size will not be
+  //  larger than 4GB).
+  return is_aligned(p, 4 * G);
```

Comment:
Does this mean that the class metadata might be late than 4 Gbyte?

Reply:
Sorry, I don't understand this question.

Reply:
Argh, typo. Does this mean that the class metadata might be more than 4G in size? Is tha the risk?

Reply:
Currently not. Class space is limited to 3Gb, CDS to 1Gb, they are placed adjacent to each other to not surpass 4G.

Reply:
So what does the word "probably" refer to?

Reply:
This function attempts to answer "is this address a valid encoding base". Strictly speaking, that question cannot be answered in isolation. The correct question "given a specific Klass range, can a valid encoding scheme be found with this base and any valid shift?".

"Probably" in this context means "This base would work, assuming a future Klass range that starts at the given encoding base and its length does not surpass the maximum encoding range".

This function is used just for one valid use case: verifying the user input for -XX:SharedBaseAddress. That happens at a point where we have not allocated anything yet, so we don't know the future encoding range.

I am really considering removing this function. It is difficult to explain, of not much use, and also slightly wrong since it ignores the base==valid EOR immediate case.

Side note about non-zero immediate + non-zero shift:

The Klass range does not have to start at the encoding base. For encoding base == zero it obviously does not. But also for non-zero encoding bases this could in theory make sense, at least on aarch64 that has no fallback add-base mode.

For example, given a Klass range starting just below 8 GB due to ASLR, this normally would lead to VM exit since such an address is unusable with MOVK, unless it fits EOR mode. But we could use 4GB as encoding base, and a non-zero shift that enlarges the encoding range, and that would still work.

We have not done this so far, and there are things to consider (e.g. it would make sense to reserve and protect a single page at encoding base, to get NULL-reference crashes), but it is possible. That is why I am hesitant to hard-code 4GB as encoding range limit.

Reply:
> This function is used just for one valid use case: verifying the
> user input for -XX:SharedBaseAddress. That happens at a point where
> we have not allocated anything yet, so we don't know the future
> encoding range.

So this function is simply asking a question that cannot be answered
without actually trying it.

> I am really considering removing this function. It is difficult to
> explain, of not much use, and also slightly wrong since it ignores
> the base==valid EOR immediate case.

Please do! If there is to be an answer about whether a base is valid,
let it happen just after a failed attempted allocation.

> Side note about non-zero immediate + non-zero shift:
> 
> The Klass range does not have to start at the encoding base. For
> encoding base == zero it obviously does not. But also for non-zero
> encoding bases this could in theory make sense, at least on aarch64
> that has no fallback add-base mode.
> 
> For example, given a Klass range starting just below 8 GB due to
> ASLR, this normally would lead to VM exit since such an address is
> unusable with MOVK, unless it fits EOR mode.

Surely we'd scan upwards until a successful allocation. Why does it
matter where the Klass range starts? Or is the problem that we're
trying hard to honour a -XX: user request, for some reason? What would
be the purpose?

Reply:
Oh, wow. I never heard about "Oracle does not want us to do that for security reasons. We are to provide some form of ASLR." Sorry, this is all news to me.